python/ray/experimental/channel.py

```diff
@@ -74,47 +101,131 @@ def __init__(
             Channel: A wrapper around ray.ObjectRef.
         """
-        if buffer_size_bytes is None:
-            if _base_ref is None:
-                raise ValueError(
-                    "One of `buffer_size_bytes` or `_base_ref` must be provided"
-                )
-            self._base_ref = _base_ref
-        else:
+        is_creator = False
+        assert len(readers) > 0
+
+        if _writer_ref is None:
             if not isinstance(buffer_size_bytes, int):
                 raise ValueError("buffer_size_bytes must be an integer")
-            self._base_ref = _create_channel_ref(buffer_size_bytes)
 
-        if not isinstance(num_readers, int):
-            raise ValueError("num_readers must be an integer")
+            self._writer_node_id = (
+                ray.runtime_context.get_runtime_context().get_node_id()
+            )
+            self._writer_ref = _create_channel_ref(self, buffer_size_bytes)
+
+            if readers[0] is None:
+                # Reader and writer are on the same node.
+                self._reader_node_id = self._writer_node_id
+                self._reader_ref = self._writer_ref
+            else:
+                # Reader and writer are on different nodes.
+                fn = readers[0].__ray_call__
+                self._reader_node_id = ray.get(fn.remote(_get_node_id))
+                if self.is_remote():
+                    self._reader_ref = ray.get(
+                        fn.remote(_create_channel_ref, buffer_size_bytes)
+                    )
+                else:
+                    self._reader_ref = self._writer_ref
+
+            is_creator = True
+        else:
+            # TODO: better error messages.
+            assert _writer_node_id is not None
+            assert _reader_ref is not None
+
+            self._writer_ref = _writer_ref
+            self._writer_node_id = _writer_node_id
+            self._reader_node_id = _reader_node_id
+            self._reader_ref = _reader_ref
+
+        self._readers = readers
+        self._buffer_size_bytes = buffer_size_bytes
 
-        self._num_readers = num_readers
         self._worker = ray._private.worker.global_worker
         self._worker.check_connected()
 
         self._writer_registered = False
         self._reader_registered = False
 
+        if is_creator:
+            self.ensure_registered_as_writer()
+            assert self._reader_ref is not None
+
+    @staticmethod
+    def is_local_node(node_id):
+        return ray.runtime_context.get_runtime_context().get_node_id() == node_id
+
+    def is_remote(self):
+        return self._writer_node_id != self._reader_node_id
+
     def ensure_registered_as_writer(self):
         if self._writer_registered:
             return
 
-        self._worker.core_worker.experimental_channel_register_writer(self._base_ref)
+        if not self.is_local_node(self._writer_node_id):
+            raise ValueError("TODO")
+
+        if self._reader_ref is None:
```

Comment:
Probably okay to just `assert self._reader_ref` here since this should be a bug.

Reply:
Done.